### PR TITLE
Spanify parts of IL reading in ILCompiler

### PR DESCRIPTION
**Pull request by PaulusParssinen**
Came across some low-hanging fruits while browsing through ILCompiler and decided to try utilize some newer .NET APIs to make the code a bit more "modern". Shouldn't be much of a performance optimization though.

### Commits
* [Spanify `ILReader`](https://github.com/dotnet/runtime/commit/a5087edc95a1c40ded5f439e95eb366962b71b5e)
  * Yet another ref struct span holder to the runtime. 
  * Make `ILReader` ref struct, only holding IL body bytes and a offset
  * Use `BinaryPrimitives` instead of manual bit-shifting
  * Remove unused `ILStreamReader.cs`
  * Make `ReadMIbcGroup` return `List<MethodProfileData>` instead of using enumerator, which prevented the use of `ILReader` (now a ref struct)
* [Spanify `ILStackHelper`](https://github.com/dotnet/runtime/commit/19ed77becfbd69bf47fd39341cb59eaca375d6dc)
  * Use `ILReader` instead to deduplicate reading logic
  * [Aand fix regression..](https://github.com/dotnet/runtime/pull/97794/commits/80b6d4b90aabdd40eefd87dbc9b21b9c1c2a7183) also combined two branches by using `ILReader.ReadBranchDestination`
* [Use built-in `Read7BitEncodedInt`](https://github.com/dotnet/runtime/commit/04138758339b51606b875aab682f7c59ba3f0514)
  * These were _recently_ exposed public.
  * This commit also removes the import `ReaderExtensions.cs` from `ILCompiler.Ryujit`. It wasn't used in the project.

Had trouble running NativeAOT smoketests locally because my hardware doesn't support AVX512, which caused `nativeaot.SmokeTests\HardwareIntrinsics\X64Avx512` to fail.

### CODE REVIEW

File: src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs [hunk 1]
```diff
@@ -25,16 +25,16 @@ public static void CheckStackBalance(this MethodIL methodIL)
...
-            int currentOffset = 0;
+            var ilReader = new ILReader(methodIL.GetILBytes());
             int stackHeight = 0;
             int maxStack = 0;
 
-            // TODO: Use Span<T> for this and stackalloc the array if reasonably sized
-            int[] stackHeights = new int[ilbytes.Length];
-            for (int i = 0; i < stackHeights.Length; i++)
-                stackHeights[i] = StackHeightNotSet;
+            Span<int> stackHeights = ilReader.Size <= StackAllocThreshold ?
```
Comment (PaulusParssinen) [inline src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs:34]: Could we get away with `short` or `ushort` representing the stack height 🤔 Quite lot wasted bytes here..

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs:34]: I think we could. The ILAsm syntax documents this as Int32, but the file format can only express 2 bytes of MaxStack for each method body.

File: src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs [hunk 2]
```diff
@@ -436,38 +400,18 @@ public static int ComputeMaxStack(this MethodIL methodIL)
...
-                        break;
-
                     default:
                         Debug.Fail("Unknown instruction");
                         break;
                 }
 
-                if (!isVariableSize)
-                    currentOffset += opcode.GetSize();
+                if (!hasReadOperand)
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs:408]: Nit: wasn't `isVariableSize` better? We defaulted to false and had to set to true in one case. Now we still default to false, but need to set it to true in two cases.

**Review by MichalStrehovsky — APPROVED** [review]
Thank you, looks great!

Cc @jakobbotsch for IBC changes.

File: src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs (hunk 2 above)
Comment (PaulusParssinen) [reply src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs:408]: I thought about name of this for a whiel. I think the meaning of the variable changed because it's using `ILReader` now.
Before we manually kept track of the `currentOffset` and `isVariableSize` was used to skip the `ILOpcode.switch_` case where the `currentOffset` was calculated manually. 
Now by using `ILReader.Skip(opcode)` (which essentially is `operandSize - op.GetSize()` + switch handling) we can avoid this manual tracking, however because we want to read the operand in two cases for stack height tracking purposes (branches and calls/newobj) we need to flip the bit to avoid "reading" the operand twice in `ILReader`. That was why I called it `hasReadOperand`. 

tldr: the current branches that use the bool are not variable sized, switch was the only special case and we now handle it in `ILReader.Skip(op)`

File: src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs (hunk 1 above)
Comment (PaulusParssinen) [reply src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs]: Hmm, [let's try it(?)](https://github.com/dotnet/runtime/pull/97794/commits/f40fcacfb2acdcaa880ad4988fc12c0f0ae21429) and hope `short.MaxValue` is enough stack for everyone. I can't find anywhere from spec whether it's signed or unsigned, I'd expect it to be read as unsigned int16 though. I wonder if some crazy generated code could hit this limit 🤔

Comment (jkotas) [reply src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs]: It is unsigned.

We have number of tests that use maxstack 65535, e.g. https://github.com/dotnet/runtime/blob/72e1ece7b683301a3ee9e0d7e8642535b4e6e89e/src/tests/JIT/Regression/JitBlue/DevDiv_359734/DevDiv_359734.il#L22

File: src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs [hunk 3]
```diff
@@ -284,12 +248,12 @@ public static int ComputeMaxStack(this MethodIL methodIL)
...
-                            MethodSignature sig = obj is MethodSignature ?
-                                (MethodSignature)obj :
-                                ((MethodDesc)obj).Signature;
-                            int adjustment = sig.Length;
+                            object obj = methodIL.GetObject(ilReader.ReadILToken());
+                            hasReadOperand = true;
+
+                            MethodSignature sig = obj is MethodSignature methodSignature ?
+                                methodSignature : ((MethodDesc)obj).Signature;
+                            short adjustment = (short)sig.Length;
```
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs:256]: There is no 16-bit limit on signature length. This cast can overflow silently.

Comment (PaulusParssinen) [reply src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs]: Came across my mind too for a moment. Reverted back to int, [reverted.](https://github.com/dotnet/runtime/pull/97794/commits/d6b2b334eb09f44f2e44e76b4157c1b003b9d98a)

File: src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs (hunk 1 above)
Comment (PaulusParssinen) [reply src/coreclr/tools/Common/TypeSystem/IL/ILStackHelper.cs:34]: Reverted in https://github.com/dotnet/runtime/pull/97794/commits/d6b2b334eb09f44f2e44e76b4157c1b003b9d98a
